# A one-way ANOVA tells us that at least two groups are different from each other. 

In [13]:
# Define the lists of data
l_1 = [7, 9, 5, 8, 6, 8, 6, 10, 7, 4]
l_2 = [4, 3, 6, 2, 7, 5, 5, 4, 1, 3]
l_3 = [6, 1, 3, 5, 3, 4, 6, 5, 7, 3]

# Calculate means
l_1_mean = sum(l_1) / len(l_1)
l_2_mean = sum(l_2) / len(l_2)
l_3_mean = sum(l_3) / len(l_3)

# Calculate overall mean
overall_mean = (l_1_mean + l_2_mean + l_3_mean) / 3

# Store means and lengths
group_means = [l_1_mean, l_2_mean, l_3_mean]
lengths = [len(l_1), len(l_2), len(l_3)]

# Print lengths, means, and overall mean
print("Lengths:", lengths)
print("Group Means:", group_means)
print("Overall Mean:", overall_mean)

# Calculate variance for each group
l_variances = [(sum((x - mean) ** 2 for x in data) / len(data)) for data, mean in zip((l_1, l_2, l_3), group_means)]
print("Variances:", l_variances)

# Calculate sum of squares between groups
ss_between = sum(length * (mean - overall_mean) ** 2 for length, mean in zip(lengths, group_means))
print("Sum of Squares Between:", ss_between)

# Calculate sum of squares within groups
ss_within = sum(variance * (length) for variance, length in zip(l_variances, lengths))
print("Sum of Squares Within:", ss_within)

# Calculate degrees of freedom
df_between = len(lengths) - 1
df_within = sum(lengths) - len(lengths)

# Calculate mean squares
ms_between = ss_between / df_between
ms_within = ss_within / df_within
print("Mean Squares Between:", ms_between)
print("Mean Squares Within:", ms_within)

# Calculate F-statistic
f_statistic = ms_between / ms_within
print("F-Statistic:", f_statistic)


Lengths: [10, 10, 10]
Group Means: [7.0, 4.0, 4.3]
Overall Mean: 5.1000000000000005
Variances: [3.0, 3.0, 3.0100000000000002]
Sum of Squares Between: 54.6
Sum of Squares Within: 90.1
Mean Squares Between: 27.3
Mean Squares Within: 3.337037037037037
F-Statistic: 8.180910099889012


# Conclusion df1 =2 df2 = 27 f-value 8.18 > 3.18 - Reject the null ho - so there u1!=u1!=u3

In [14]:
r = ss_between/ss_within
print(r)

0.6059933407325194


Hence 60% of the difference between the scores is because of the approach that was used. Rest 40% is unknown. Hence, the Eta square helps us conclude whether the independent variable really impacts the dependent variable or whether the difference is due to chance or any other factor.